src/vm.c

```diff
@@ -203,6 +206,28 @@ int vm_load_initrd(vm_t *v, const char *initrd_path)
     return 0;
 }
 
+void vm_handle_io(vm_t *v, struct kvm_run *run)
+{
+    uint64_t addr = run->io.port;
+    void *data = (void *) run + run->io.data_offset;
+    bool is_write = run->io.direction == KVM_EXIT_IO_OUT;
+
+    if (run->io.port >= COM1_PORT_BASE && run->io.port < COM1_PORT_END) {
+        serial_handle(&v->serial, run);
+    } else {
+        for (int i = 0; i < run->io.count; i++) {
+            bus_handle_io(&v->io_bus, data, is_write, addr, run->io.size);
```

Comment:
Is it make sense to also register serial UART  as an `io_bus` object? So the emulated hardware with read and write method could be implemented on the same abstraction.

Reply:
@RinHizakura, I tend to merge this pull request and then think of further abstractions. Can you create a new issue which consolidates serial UART refactoring?